File: sdks/typescript/src/parser.ts
/**
 * Parser for CrowdControl policy source.
 *
 * Ports github.com/mikemackintosh/crowdcontrol/parser/parser.go to TypeScript.
 */

import {
  LexError,
  Token,
  TokenType,
  lex,
  tokenDisplay,
  tokenName,
} from "./lexer.js";
import {
  Condition,
  ConditionType,
  Expr,
  ExprKind,
  Policy,
  Rule,
  newCondition,
  newExpr,
  newPolicy,
  newRule,
} from "./types.js";

export class ParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ParseError";
  }
}

/** Parse CrowdControl source into a Policy AST. */
export function parse(source: string): Policy {
  let tokens: Token[];
  try {
    tokens = lex(source);
  } catch (e) {
    if (e instanceof LexError) {
      throw new ParseError(e.message);
    }
    throw e;
  }
  const p = new Parser(tokens);
  return p.parsePolicy();
}

const COMPARISON_TOKENS = new Set<TokenType>([
  TokenType.EQ,
  TokenType.NEQ,
  TokenType.LT,
  TokenType.GT,
  TokenType.LTE,
  TokenType.GTE,
]);

const ARITH_TOKENS = new Set<TokenType>([
  TokenType.PLUS,
  TokenType.MINUS,
  TokenType.STAR,
  TokenType.SLASH,
]);

class Parser {
  private tokens: Token[];
  private pos = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  // ----- top level -------------------------------------------------------

  parsePolicy(): Policy {
    const policy = newPolicy();
    while (!this.atEnd()) {
      policy.rules.push(this.parseRule());
    }
    return policy;
  }

  private parseRule(): Rule {
    const kindTok = this.advance();
    if (kindTok.type !== TokenType.IDENT) {
      throw this.errorf(`expected forbid, warn, or permit, got ${tokenDisplay(kindTok)}`);
    }
    const kind = kindTok.val;
    if (kind !== "forbid" && kind !== "warn" && kind !== "permit") {
      throw this.errorf(`expected forbid, warn, or permit, got ${JSON.stringify(kind)}`);
    }

    const nameTok = this.advance();
    if (nameTok.type !== TokenType.STRING) {
      throw this.errorf(`expected rule name string, got ${tokenDisplay(nameTok)}`);
    }

    this.expect(TokenType.LBRACE);

    const rule = newRule({ kind, name: nameTok.val });

    while (!this.check(TokenType.RBRACE) && !this.atEnd()) {
      this.parseClause(rule);
    }

    this.expect(TokenType.RBRACE);
    return rule;
  }

  private parseClause(rule: Rule): void {
    const tok = this.peek();
    const val = tok.val;

    if (val === "unless") {
      this.parseUnless(rule);
      return;
    }
    if (val === "message") {
      this.parseMessage(rule);
      return;
    }
    if (val === "description" || val === "owner" || val === "link") {
      this.parseMetadata(rule, val);
      return;
    }

    let cond: Condition;
    if (val === "not") {
      cond = this.parseNegatedCondition();
    } else if (val === "any" || val === "all") {
      cond = this.parseQuantifier();
    } else if (val === "has") {
      cond = this.parseHasCondition();
    } else if (val === "count") {
      cond = this.parseAggregateCondition();
    } else if (val === "lower" || val === "upper" || val === "len") {
      cond = this.parseTransformCondition();
    } else {
      cond = this.parseFieldCond();
    }

    rule.conditions.push(this.wrapOr(cond));
  }

  // ----- or-chaining -----------------------------------------------------

  private wrapOr(first: Condition): Condition {
    if (!this.checkIdent("or")) {
      return first;
    }
    const group: Condition[] = [first];
    while (this.checkIdent("or")) {
      this.advance(); // consume "or"
      try {
        const cond = this.parseSingleCondition();
        group.push(cond);
      } catch (e) {
        if (e instanceof ParseError) {
          break;
        }
        throw e;
      }
    }
    return newCondition({ type: ConditionType.OR, orGroup: group });
  }

  private parseSingleCondition(): Condition {
    const val = this.peek().val;
    if (val === "not") return this.parseNegatedCondition();
    if (val === "any" || val === "all") return this.parseQuantifier();
    if (val === "has") return this.parseHasCondition();
    if (val === "count") return this.parseAggregateCondition();
    if (val === "lower" || val === "upper" || val === "len") {
      return this.parseTransformCondition();
    }
    return this.parseFieldCond();
  }

  // ----- individual conditions -------------------------------------------

  private parseNegatedCondition(): Condition {
    this.advance(); // consume "not"
    const cond = this.parseSingleCondition();
    cond.negated = !cond.negated;
    return cond;
  }

  private parseHasCondition(): Condition {
    this.advance(); // consume "has"
    const field = this.parseDottedPath();
    return newCondition({ type: ConditionType.HAS, field });
  }

  private parseQuantifier(): Condition {
    const quantTok = this.advance(); // "any" or "all"
    const listField = this.parseDottedPath();
    const predicate = this.parseElementPredicate();
    const condType =
      quantTok.val === "all" ? ConditionType.ALL : ConditionType.ANY;
    return newCondition({
      type: condType,
      quantifier: quantTok.val,
      listField,
      predicate,
    });
  }

  private parseElementPredicate(): Condition {
    const tok = this.peek();

    if (tok.type === TokenType.IDENT && tok.val === "in") {
      this.advance();
      const nxt = this.peek();
      if (nxt.type === TokenType.LBRACKET) {
        const vals = this.parseStringList();
        return newCondition({ type: ConditionType.FIELD, op: "in", value: vals });
      }
      throw this.errorf(`expected list after 'in', got ${tokenDisplay(nxt)}`);
    }

    if (
      tok.type === TokenType.IDENT &&
      (tok.val === "matches" || tok.val === "matches_regex")
    ) {
      const op = tok.val;
      this.advance();
      const valTok = this.advance();
      if (valTok.type !== TokenType.STRING) {
        throw this.errorf(`${op} expects a string pattern, got ${tokenDisplay(valTok)}`);
      }
      return newCondition({ type: ConditionType.FIELD, op, value: valTok.val });
    }

    if (tok.type === TokenType.IDENT && tok.val === "contains") {
      this.advance();
      const val = this.parseValue();
      return newCondition({ type: ConditionType.FIELD, op: "contains", value: val });
    }

    const opTok = this.advance();
    if (!COMPARISON_TOKENS.has(opTok.type)) {
      throw this.errorf(
        `expected operator in quantifier predicate, got ${tokenDisplay(opTok)}`,
      );
    }
    const val = this.parseValue();
    return newCondition({ type: ConditionType.FIELD, op: opTok.val, value: val });
  }

  private parseFieldCond(): Condition {
    const field = this.parseDottedPath();

    if (this.isArithOp()) {
      const left = newExpr({ kind: ExprKind.FIELD, field });
      return this.parseExprConditionFromLeft(left);
    }

    const opTok = this.advance();
    let op = opTok.val;

    if (!COMPARISON_TOKENS.has(opTok.type)) {
      if (
        opTok.type === TokenType.IDENT &&
        (op === "in" ||
          op === "matches" ||
          op === "matches_regex" ||
          op === "contains" ||
          op === "intersects" ||
          op === "is_subset")
      ) {
        // keep op
      } else {
        throw this.errorf(`expected operator, got ${tokenDisplay(opTok)}`);
      }
    }

    const cond = newCondition({ type: ConditionType.FIELD, field, op });

    if (op === "in" || op === "intersects" || op === "is_subset") {
      cond.value = this.parseStringList();
    } else if (op === "matches" || op === "matches_regex") {
      const valTok = this.advance();
      if (valTok.type !== TokenType.STRING) {
        throw this.errorf(`${op} expects a string pattern, got ${tokenDisplay(valTok)}`);
      }
      cond.value = valTok.val;
    } else {
      cond.value = this.parseValue();
    }

    return cond;
  }

  private isArithOp(): boolean {
    return ARITH_TOKENS.has(this.peek().type);
  }

  private parseExprConditionFromLeft(left: Expr): Condition {
    const leftExpr = this.parseArithExprFrom(left);

    const opTok = this.advance();
    if (!COMPARISON_TOKENS.has(opTok.type)) {
      throw this.errorf(
        `expected comparison operator in expression, got ${tokenDisplay(opTok)}`,
      );
    }

    const rightExpr = this.parseArithExpr();
    return newCondition({
      type: ConditionType.EXPR,
      op: opTok.val,
      leftExpr,
      rightExpr,
    });
  }

  private parseArithExpr(): Expr {
    const left = this.parseExprTerm();
    return this.parseArithExprFrom(left);
  }

  private parseArithExprFrom(left: Expr): Expr {
    let current = left;
    while (this.isArithOp()) {
      const opTok = this.advance();
      const right = this.parseExprTerm();
      current = newExpr({
        kind: ExprKind.BINARY,
        op: opTok.val,
        left: current,
        right,
      });
    }
    return current;
  }

  private parseExprTerm(): Expr {
    const tok = this.peek();

    if (tok.type === TokenType.NUMBER) {
      this.advance();
      const num = Number(tok.val);
      if (Number.isNaN(num)) {
        throw this.errorf(`invalid number: ${tok.val}`);
      }
      return newExpr({ kind: ExprKind.LITERAL, value: num });
    }

    if (tok.type === TokenType.IDENT && (tok.val === "count" || tok.val === "len")) {
      const funcName = tok.val;
      this.advance();
      this.expect(TokenType.LPAREN);
      const path = this.parseDottedPath();
      this.expect(TokenType.RPAREN);
      if (funcName === "count") {
        return newExpr({ kind: ExprKind.COUNT, aggTarget: path });
      }
      return newExpr({ kind: ExprKind.LEN, field: path, transform: "len" });
    }

    if (tok.type === TokenType.IDENT) {
      const path = this.parseDottedPath();
      return newExpr({ kind: ExprKind.FIELD, field: path });
    }

    throw this.errorf(
      `expected number, field, count(), or len() in expression, got ${tokenDisplay(tok)}`,
    );
  }

  private parseUnless(rule: Rule): void {
    this.advance(); // consume "unless"
    const val = this.peek().val;
    let cond: Condition;
    if (val === "not") {
      cond = this.parseNegatedCondition();
    } else if (val === "any" || val === "all") {
      cond = this.parseQuantifier();
    } else if (val === "has") {
      cond = this.parseHasCondition();
    } else if (val === "lower" || val === "upper" || val === "len") {
      cond = this.parseTransformCondition();
    } else if (val === "count") {
      cond = this.parseAggregateCondition();
    } else {
      cond = this.parseFieldCond();
    }
    rule.unlesses.push(cond);
  }

  private parseTransformCondition(): Condition {
    const funcTok = this.advance();
    this.expect(TokenType.LPAREN);
    const field = this.parseDottedPath();
    this.expect(TokenType.RPAREN);

    if (funcTok.val === "len" && this.isArithOp()) {
      const left = newExpr({ kind: ExprKind.LEN, field, transform: "len" });
      return this.parseExprConditionFromLeft(left);
    }

    const opTok = this.advance();
    let op = opTok.val;
    if (!COMPARISON_TOKENS.has(opTok.type)) {
      if (
        opTok.type === TokenType.IDENT &&
        (op === "in" || op === "matches" || op === "matches_regex" || op === "contains")
      ) {
        // ok
      } else {
        throw this.errorf(
          `expected operator after ${funcTok.val}(), got ${tokenDisplay(opTok)}`,
        );
      }
    }

    const cond = newCondition({
      type: ConditionType.FIELD,
      field,
      op,
      transform: funcTok.val,
    });

    if (op === "in") {
      cond.value = this.parseStringList();
    } else if (op === "matches" || op === "matches_regex") {
      const valTok = this.advance();
      if (valTok.type !== TokenType.STRING) {
        throw this.errorf(`${op} expects a string pattern, got ${tokenDisplay(valTok)}`);
      }
      cond.value = valTok.val;
    } else {
      cond.value = this.parseValue();
    }
    return cond;
  }

  private parseAggregateCondition(): Condition {
    this.advance(); // consume "count"
    this.expect(TokenType.LPAREN);
    const target = this.parseDottedPath();
    this.expect(TokenType.RPAREN);

    if (this.isArithOp()) {
      const left = newExpr({ kind: ExprKind.COUNT, aggTarget: target });
      return this.parseExprConditionFromLeft(left);
    }

    const opTok = this.advance();
    if (!COMPARISON_TOKENS.has(opTok.type)) {
      throw this.errorf(
        `expected comparison operator after count(), got ${tokenDisplay(opTok)}`,
      );
    }

    const valTok = this.advance();
    if (valTok.type !== TokenType.NUMBER) {
      throw this.errorf(`expected number after operator, got ${tokenDisplay(valTok)}`);
    }
    const num = Number.parseInt(valTok.val, 10);
    if (Number.isNaN(num)) {
      throw this.errorf(`invalid number: ${valTok.val}`);
    }

    return newCondition({
      type: ConditionType.AGGREGATE,
      aggregateFunc: "count",
      aggregateTarget: target,
      op: opTok.val,
      value: num,
    });
  }

  private parseMessage(rule: Rule): void {
    this.advance();
    const msgTok = this.advance();
    if (msgTok.type !== TokenType.STRING) {
      throw this.errorf(`expected message string, got ${tokenDisplay(msgTok)}`);
    }
    rule.message = msgTok.val;
  }

  private parseMetadata(rule: Rule, keyword: string): void {
    this.advance();
    const valTok = this.advance();
    if (valTok.type !== TokenType.STRING) {
      throw this.errorf(`expected string after ${keyword}, got ${tokenDisplay(valTok)}`);
    }
    if (keyword === "description") rule.description = valTok.val;
    else if (keyword === "owner") rule.owner = valTok.val;
    else if (keyword === "link") rule.link = valTok.val;
  }

  // ----- helpers ---------------------------------------------------------

  private parseDottedPath(): string {
    const tok = this.advance();
    if (tok.type !== TokenType.IDENT) {
      throw this.errorf(`expected identifier, got ${tokenDisplay(tok)}`);
    }
    const parts: string[] = [tok.val];
    while (this.check(TokenType.DOT)) {
      this.advance();
      const nxt = this.advance();
      if (nxt.type !== TokenType.IDENT) {
        throw this.errorf(`expected identifier after '.', got ${tokenDisplay(nxt)}`);
      }
      parts.push(nxt.val);
    }
    return parts.join(".");
  }

  private parseStringList(): string[] {
    this.expect(TokenType.LBRACKET);
    const vals: string[] = [];
    while (!this.check(TokenType.RBRACKET) && !this.atEnd()) {
      const tok = this.advance();
      if (tok.type !== TokenType.STRING) {
        throw this.errorf(`expected string in list, got ${tokenDisplay(tok)}`);
      }
      vals.push(tok.val);
      if (this.check(TokenType.COMMA)) {
        this.advance();
      }
    }
    this.expect(TokenType.RBRACKET);
    return vals;
  }

  private parseValue(): unknown {
    const tok = this.advance();
    if (tok.type === TokenType.STRING) return tok.val;
    if (tok.type === TokenType.NUMBER) {
      // Preserve integer-vs-float the way the Python port does: ints parse
      // to JS number (integer), floats also parse to number. Aggregate
      // count() target will re-parse as int. For equality via _fmtV, ints
      // print without decimal which is what we want.
      const num = Number(tok.val);
      if (Number.isNaN(num)) {
        throw this.errorf(`invalid number: ${tok.val}`);
      }
      return num;
    }
    if (tok.type === TokenType.IDENT) {
      if (tok.val === "true") return true;
      if (tok.val === "false") return false;
      throw this.errorf(`unexpected identifier ${JSON.stringify(tok.val)} in value position`);
    }
    throw this.errorf(`expected value, got ${tokenDisplay(tok)}`);
  }

  private peek(): Token {
    if (this.pos >= this.tokens.length) {
      return { type: TokenType.EOF, val: "", line: 0, col: 0 };
    }
    return this.tokens[this.pos]!;
  }

  private advance(): Token {
    const tok = this.peek();
    if (tok.type !== TokenType.EOF) {
      this.pos++;
    }
    return tok;
  }

  private check(t: TokenType): boolean {
    return this.peek().type === t;
  }

  private checkIdent(val: string): boolean {
    const tok = this.peek();
    return tok.type === TokenType.IDENT && tok.val === val;
  }

  private atEnd(): boolean {
    return this.peek().type === TokenType.EOF;
  }

  private expect(t: TokenType): Token {
    const tok = this.advance();
    if (tok.type !== t) {
      throw this.errorf(`expected ${tokenName(t)}, got ${tokenDisplay(tok)}`);
    }
    return tok;
  }

  private errorf(msg: string): ParseError {
    const tok = this.peek();
    return new ParseError(`line ${tok.line} col ${tok.col}: ${msg}`);
  }
}

File: sdks/kotlin/src/main/kotlin/io/github/mikemackintosh/crowdcontrol/Parser.kt
/*
 * Parser for CrowdControl policy source.
 *
 * Ports github.com/mikemackintosh/crowdcontrol/parser/parser.go to Kotlin.
 */
package io.github.mikemackintosh.crowdcontrol

/** Parse CrowdControl source into a Policy AST. */
fun parse(source: String): Policy {
    val tokens = lex(source)
    return Parser(tokens).parsePolicy()
}

internal class Parser(private val tokens: List<Token>) {
    private var pos: Int = 0

    // ----- top level -------------------------------------------------------

    fun parsePolicy(): Policy {
        val policy = Policy()
        while (!atEnd()) {
            policy.rules.add(parseRule())
        }
        return policy
    }

    private fun parseRule(): Rule {
        val kindTok = advance()
        if (kindTok.type != TokenType.IDENT) {
            throw errorf("expected forbid, warn, or permit, got $kindTok")
        }
        val kind = kindTok.value
        if (kind != "forbid" && kind != "warn" && kind != "permit") {
            throw errorf("expected forbid, warn, or permit, got \"$kind\"")
        }

        val nameTok = advance()
        if (nameTok.type != TokenType.STRING) {
            throw errorf("expected rule name string, got $nameTok")
        }

        expect(TokenType.LBRACE)

        val rule = Rule(kind = kind, name = nameTok.value)

        while (!check(TokenType.RBRACE) && !atEnd()) {
            parseClause(rule)
        }

        expect(TokenType.RBRACE)
        return rule
    }

    private fun parseClause(rule: Rule) {
        val tok = peek()
        val v = tok.value

        if (v == "unless") {
            parseUnless(rule)
            return
        }
        if (v == "message") {
            parseMessage(rule)
            return
        }
        if (v == "description" || v == "owner" || v == "link") {
            parseMetadata(rule, v)
            return
        }
        val cond: Condition = when (v) {
            "not" -> parseNegatedCondition()
            "any", "all" -> parseQuantifier()
            "has" -> parseHasCondition()
            "count" -> parseAggregateCondition()
            "lower", "upper", "len" -> parseTransformCondition()
            else -> parseFieldCond()
        }

        rule.conditions.add(wrapOr(cond))
    }

    // ----- or-chaining -----------------------------------------------------

    private fun wrapOr(first: Condition): Condition {
        if (!checkIdent("or")) return first
        val group = mutableListOf(first)
        while (checkIdent("or")) {
            advance() // consume "or"
            val cond = try {
                parseSingleCondition()
            } catch (_: ParseError) {
                break
            }
            group.add(cond)
        }
        return Condition(type = ConditionType.OR, orGroup = group)
    }

    private fun parseSingleCondition(): Condition {
        val v = peek().value
        return when (v) {
            "not" -> parseNegatedCondition()
            "any", "all" -> parseQuantifier()
            "has" -> parseHasCondition()
            "count" -> parseAggregateCondition()
            "lower", "upper", "len" -> parseTransformCondition()
            else -> parseFieldCond()
        }
    }

    // ----- individual conditions -------------------------------------------

    private fun parseNegatedCondition(): Condition {
        advance() // consume "not"
        val cond = parseSingleCondition()
        cond.negated = !cond.negated
        return cond
    }

    private fun parseHasCondition(): Condition {
        advance() // consume "has"
        val field = parseDottedPath()
        return Condition(type = ConditionType.HAS, field = field)
    }

    private fun parseQuantifier(): Condition {
        val quantTok = advance() // consume "any" or "all"
        val listField = parseDottedPath()
        val predicate = parseElementPredicate()
        val condType = if (quantTok.value == "all") ConditionType.ALL else ConditionType.ANY
        return Condition(
            type = condType,
            quantifier = quantTok.value,
            listField = listField,
            predicate = predicate,
        )
    }

    private fun parseElementPredicate(): Condition {
        val tok = peek()

        if (tok.type == TokenType.IDENT && tok.value == "in") {
            advance()
            val nxt = peek()
            if (nxt.type == TokenType.LBRACKET) {
                val vals = parseStringList()
                return Condition(type = ConditionType.FIELD, op = "in", value = vals)
            }
            throw errorf("expected list after 'in', got $nxt")
        }

        if (tok.type == TokenType.IDENT && (tok.value == "matches" || tok.value == "matches_regex")) {
            val op = tok.value
            advance()
            val valTok = advance()
            if (valTok.type != TokenType.STRING) {
                throw errorf("$op expects a string pattern, got $valTok")
            }
            return Condition(type = ConditionType.FIELD, op = op, value = valTok.value)
        }

        if (tok.type == TokenType.IDENT && tok.value == "contains") {
            advance()
            val v = parseValue()
            return Condition(type = ConditionType.FIELD, op = "contains", value = v)
        }

        val opTok = advance()
        if (opTok.type !in COMPARISON_TOKENS) {
            throw errorf("expected operator in quantifier predicate, got $opTok")
        }
        val v = parseValue()
        return Condition(type = ConditionType.FIELD, op = opTok.value, value = v)
    }

    private fun parseFieldCond(): Condition {
        val field = parseDottedPath()

        if (isArithOp()) {
            val left = Expr(kind = ExprKind.FIELD, field = field)
            return parseExprConditionFromLeft(left)
        }

        val opTok = advance()
        var op = opTok.value

        if (opTok.type !in COMPARISON_TOKENS) {
            if (opTok.type == TokenType.IDENT && opTok.value in KEYWORD_OPS) {
                op = opTok.value
            } else {
                throw errorf("expected operator, got $opTok")
            }
        }

        val cond = Condition(type = ConditionType.FIELD, field = field, op = op)

        when (op) {
            "in", "intersects", "is_subset" -> cond.value = parseStringList()
            "matches", "matches_regex" -> {
                val valTok = advance()
                if (valTok.type != TokenType.STRING) {
                    throw errorf("$op expects a string pattern, got $valTok")
                }
                cond.value = valTok.value
            }
            else -> cond.value = parseValue()
        }

        return cond
    }

    private fun isArithOp(): Boolean {
        val t = peek().type
        return t == TokenType.PLUS || t == TokenType.MINUS ||
            t == TokenType.STAR || t == TokenType.SLASH
    }

    private fun parseExprConditionFromLeft(left: Expr): Condition {
        val leftExpr = parseArithExprFrom(left)

        val opTok = advance()
        if (opTok.type !in COMPARISON_TOKENS) {
            throw errorf("expected comparison operator in expression, got $opTok")
        }

        val rightExpr = parseArithExpr()
        return Condition(
            type = ConditionType.EXPR,
            op = opTok.value,
            leftExpr = leftExpr,
            rightExpr = rightExpr,
        )
    }

    private fun parseArithExpr(): Expr {
        val left = parseExprTerm()
        return parseArithExprFrom(left)
    }

    private fun parseArithExprFrom(initial: Expr): Expr {
        var left = initial
        while (isArithOp()) {
            val opTok = advance()
            val right = parseExprTerm()
            left = Expr(kind = ExprKind.BINARY, op = opTok.value, left = left, right = right)
        }
        return left
    }

    private fun parseExprTerm(): Expr {
        val tok = peek()

        if (tok.type == TokenType.NUMBER) {
            advance()
            val num = tok.value.toDoubleOrNull()
                ?: throw errorf("invalid number: ${tok.value}")
            return Expr(kind = ExprKind.LITERAL, value = num)
        }

        if (tok.type == TokenType.IDENT && (tok.value == "count" || tok.value == "len")) {
            val funcName = tok.value
            advance()
            expect(TokenType.LPAREN)
            val path = parseDottedPath()
            expect(TokenType.RPAREN)
            return if (funcName == "count") {
                Expr(kind = ExprKind.COUNT, aggTarget = path)
            } else {
                Expr(kind = ExprKind.LEN, field = path, transform = "len")
            }
        }

        if (tok.type == TokenType.IDENT) {
            val path = parseDottedPath()
            return Expr(kind = ExprKind.FIELD, field = path)
        }

        throw errorf("expected number, field, count(), or len() in expression, got $tok")
    }

    private fun parseUnless(rule: Rule) {
        advance() // consume "unless"
        val v = peek().value
        val cond: Condition = when (v) {
            "not" -> parseNegatedCondition()
            "any", "all" -> parseQuantifier()
            "has" -> parseHasCondition()
            "lower", "upper", "len" -> parseTransformCondition()
            "count" -> parseAggregateCondition()
            else -> parseFieldCond()
        }
        rule.unlesses.add(cond)
    }

    private fun parseTransformCondition(): Condition {
        val funcTok = advance()
        expect(TokenType.LPAREN)
        val field = parseDottedPath()
        expect(TokenType.RPAREN)

        if (funcTok.value == "len" && isArithOp()) {
            val left = Expr(kind = ExprKind.LEN, field = field, transform = "len")
            return parseExprConditionFromLeft(left)
        }

        val opTok = advance()
        var op = opTok.value
        if (opTok.type !in COMPARISON_TOKENS) {
            if (opTok.type == TokenType.IDENT && opTok.value in setOf("in", "matches", "matches_regex", "contains")) {
                op = opTok.value
            } else {
                throw errorf("expected operator after ${funcTok.value}(), got $opTok")
            }
        }

        val cond = Condition(type = ConditionType.FIELD, field = field, op = op, transform = funcTok.value)

        when (op) {
            "in" -> cond.value = parseStringList()
            "matches", "matches_regex" -> {
                val valTok = advance()
                if (valTok.type != TokenType.STRING) {
                    throw errorf("$op expects a string pattern, got $valTok")
                }
                cond.value = valTok.value
            }
            else -> cond.value = parseValue()
        }
        return cond
    }

    private fun parseAggregateCondition(): Condition {
        advance() // consume "count"
        expect(TokenType.LPAREN)
        val target = parseDottedPath()
        expect(TokenType.RPAREN)

        if (isArithOp()) {
            val left = Expr(kind = ExprKind.COUNT, aggTarget = target)
            return parseExprConditionFromLeft(left)
        }

        val opTok = advance()
        if (opTok.type !in COMPARISON_TOKENS) {
            throw errorf("expected comparison operator after count(), got $opTok")
        }

        val valTok = advance()
        if (valTok.type != TokenType.NUMBER) {
            throw errorf("expected number after operator, got $valTok")
        }
        val num = valTok.value.toIntOrNull()
            ?: throw errorf("invalid number: ${valTok.value}")

        return Condition(
            type = ConditionType.AGGREGATE,
            aggregateFunc = "count",
            aggregateTarget = target,
            op = opTok.value,
            value = num,
        )
    }

    private fun parseMessage(rule: Rule) {
        advance()
        val msgTok = advance()
        if (msgTok.type != TokenType.STRING) {
            throw errorf("expected message string, got $msgTok")
        }
        rule.message = msgTok.value
    }

    private fun parseMetadata(rule: Rule, keyword: String) {
        advance()
        val valTok = advance()
        if (valTok.type != TokenType.STRING) {
            throw errorf("expected string after $keyword, got $valTok")
        }
        when (keyword) {
            "description" -> rule.description = valTok.value
            "owner" -> rule.owner = valTok.value
            "link" -> rule.link = valTok.value
        }
    }

    // ----- helpers ---------------------------------------------------------

    private fun parseDottedPath(): String {
        val tok = advance()
        if (tok.type != TokenType.IDENT) {
            throw errorf("expected identifier, got $tok")
        }
        val parts = mutableListOf(tok.value)
        while (check(TokenType.DOT)) {
            advance()
            val nxt = advance()
            if (nxt.type != TokenType.IDENT) {
                throw errorf("expected identifier after '.', got $nxt")
            }
            parts.add(nxt.value)
        }
        return parts.joinToString(".")
    }

    private fun parseStringList(): List<String> {
        expect(TokenType.LBRACKET)
        val vals = mutableListOf<String>()
        while (!check(TokenType.RBRACKET) && !atEnd()) {
            val tok = advance()
            if (tok.type != TokenType.STRING) {
                throw errorf("expected string in list, got $tok")
            }
            vals.add(tok.value)
            if (check(TokenType.COMMA)) advance()
        }
        expect(TokenType.RBRACKET)
        return vals
    }

    private fun parseValue(): Any {
        val tok = advance()
        when (tok.type) {
            TokenType.STRING -> return tok.value
            TokenType.NUMBER -> {
                // Prefer Long/Int for integer-valued numbers, Double otherwise.
                return if ('.' in tok.value) {
                    tok.value.toDoubleOrNull()
                        ?: throw errorf("invalid number: ${tok.value}")
                } else {
                    tok.value.toLongOrNull()?.let { l ->
                        // Fit in Int if possible for parity with other ports.
                        if (l in Int.MIN_VALUE.toLong()..Int.MAX_VALUE.toLong()) l.toInt() else l
                    } ?: tok.value.toDoubleOrNull()
                        ?: throw errorf("invalid number: ${tok.value}")
                }
            }
            TokenType.IDENT -> {
                return when (tok.value) {
                    "true" -> true
                    "false" -> false
                    else -> throw errorf("unexpected identifier \"${tok.value}\" in value position")
                }
            }
            else -> throw errorf("expected value, got $tok")
        }
    }

    private fun peek(): Token {
        if (pos >= tokens.size) return Token(TokenType.EOF)
        return tokens[pos]
    }

    private fun advance(): Token {
        val tok = peek()
        if (tok.type != TokenType.EOF) pos += 1
        return tok
    }

    private fun check(t: TokenType): Boolean = peek().type == t

    private fun checkIdent(v: String): Boolean {
        val tok = peek()
        return tok.type == TokenType.IDENT && tok.value == v
    }

    private fun atEnd(): Boolean = peek().type == TokenType.EOF

    private fun expect(t: TokenType): Token {
        val tok = advance()
        if (tok.type != t) throw errorf("expected ${t.name}, got $tok")
        return tok
    }

    private fun errorf(msg: String): ParseError {
        val tok = peek()
        return ParseError("line ${tok.line} col ${tok.col}: $msg")
    }

    companion object {
        private val COMPARISON_TOKENS = setOf(
            TokenType.EQ, TokenType.NEQ,
            TokenType.LT, TokenType.GT,
            TokenType.LTE, TokenType.GTE,
        )
        private val KEYWORD_OPS = setOf(
            "in", "matches", "matches_regex",
            "contains", "intersects", "is_subset",
        )
    }
}
